### PR TITLE
MySqlCluster dump based replication

### PR DESCRIPTION
**Pull request by neykov**
When the binary log of the mysql master is truncated the only way to spin up new slaves is to do a dump of the master and initialize the slave from it - implemented here.

### CODE REVIEW

File: software/database/src/main/java/org/apache/brooklyn/entity/database/mysql/InitSlaveTaskBody.java [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+    private MySqlNode getRandomSlave() {
+        List<MySqlNode> slaves = getHealhtySlaves();
+        if (slaves.size() > 0) {
+            return slaves.get(new Random().nextInt(slaves.size()));
+        } else {
+            return getMaster();
+        }
+    }
+
+    private ImmutableList<MySqlNode> getHealhtySlaves() {
```
Comment (sjcorbett) [inline software/database/src/main/java/org/apache/brooklyn/entity/database/mysql/InitSlaveTaskBody.java]: Typo in method name: `Healhty`

File: utils/common/src/main/java/org/apache/brooklyn/util/GenericTypes.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.reflect.TypeToken;
+
+@SuppressWarnings("serial")
+public class GenericTypes {
```
Comment (sjcorbett) [inline utils/common/src/main/java/org/apache/brooklyn/util/GenericTypes.java]: Add a private constructor.

**Comment by sjcorbett** [comment]
Looks great. I've made two trivial comments. Happy for this to be merged.

File: utils/common/src/main/java/org/apache/brooklyn/util/GenericTypes.java (hunk 2 above)
Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/GenericTypes.java]: Converted to an interface instead, allows to remove all `public static final` declarations.

**Comment by neykov** [comment]
Thanks @sjcorbett, addressed comments.

**Comment by iyovcheva** [comment]
There is an error after last commit:
[ERROR] /incubator-brooklyn/software/database/src/main/java/org/apache/brooklyn/entity/database/mysql/MySqlClusterImpl.java:[124,17] fromMembers() in org.apache.brooklyn.enricher.stock.Enrichers.AbstractAggregatorBuilder is defined in an inaccessible class or interface
[ERROR] /incubator-brooklyn/software/database/src/main/java/org/apache/brooklyn/entity/database/mysql/MySqlClusterImpl.java:[125,17] build() in org.apache.brooklyn.enricher.stock.Enrichers.AbstractAggregatorBuilder is defined in an inaccessible class or interface

**Comment by neykov** [comment]
Thanks @iyovcheva, seems to be a problem with Java 8 only, working on it.

**Comment by neykov** [comment]
Java 8 build fixed in https://github.com/apache/incubator-brooklyn/pull/991

**Pull request closed without merging**